### PR TITLE
Program analysis

### PR DESCRIPTION
**Pull request by RamIowaStateUniversity**
Added traversal strategies, getValue function, test-traversal.boa, added constructor and clone method to tuple type.

### CODE REVIEW

File: src/java/boa/compiler/visitors/TypeCheckingVisitor.java [hunk 1]
```diff
@@ -487,6 +489,17 @@ public void visit(final Factor n, final SymbolTable env) {
 						throw new TypeCheckException(n.getOperand(), e.getMessage(), e);
 					}
 					type = v.getFunction().erase(formalParameters);
+					if(((Identifier)n.getOperand()).getToken().equals("getValue")) {
```
Comment (psybers) [inline src/java/boa/compiler/visitors/TypeCheckingVisitor.java]: This feels like a nasty hack.  What is the purpose of these changes to the type checker?

**Comment by RamIowaStateUniversity** [comment]
The return type of getValue changes based on the traversal passed to it.
Therefore when we encounter getValue, we look at the traversal passed to it
and assign the return type of the traversal as the return type of getValue.
If getValue has the current traversal passed to it (ie when we dont pass
any traversal to getValue, in other word if the size of the arguments
passed is one), then we assign the last seen return type(which is the
return type of current traversal) as the type of getValue.

**Comment by psybers** [comment]
That first case is why we have BoaTypeVar in the compiler.  Take a look at other functions dealing with sets/stacks/etc such as keys() and values().  These return a type depending on the type given as argument. This is the proper way to handle that.

**Comment by hridesh** [comment]
Thanks for all of your work on this. I have two suggestions for the test for traversal. 
1. The test should be split into smaller tests that make it possible to identify which features were affected by the change.
2. More combinations should be exercised in the test case.  Currently, not all permutations and combinations of traversal and traverse are used by test cases.

**Comment by RamIowaStateUniversity** [comment]
Dr.Dyer,

I modified getValue to use the passed traversals return type through
TypeVar. But in the other case, where the traversal is not passed and the
current traversal is used, this is more like a recursive call which we
don't support and therefore in this case, I do have to do some kind of a
hack to assign type.

I also made another commit where I added enhanced for loop to support Set
iteration. With this commit, the program-analysis feature is done. After
merging this, I will make one more pull request with more test cases as per
Dr.Rajan's suggesstion and then move on to adding hybrid strategy.

Thanks,
Ram

File: src/java/boa/compiler/SymbolTable.java [hunk 2]
```diff
@@ -188,7 +193,10 @@
 		// set functions
 		globalFunctions.addFunction("contains", new BoaFunction(new BoaBool(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaTypeVar("V") }, "${0}.contains(${1})"));
 		globalFunctions.addFunction("add", new BoaFunction(new BoaAny(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaTypeVar("V") }, "${0}.add(${1})"));
+		globalFunctions.addFunction("addAll", new BoaFunction(new BoaAny(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaSet(new BoaTypeVar("V")) }, "${0}.addAll(${1})"));
```
Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: isnt this the same as the already existing union() a few lines below?

File: src/java/boa/compiler/SymbolTable.java [hunk 3]
```diff
@@ -188,7 +193,10 @@
...
+		globalFunctions.addFunction("containsAll", new BoaFunction(new BoaBool(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaSet(new BoaTypeVar("V"))}, "${0}.containsAll(${1})"));
 		globalFunctions.addFunction("remove", new BoaFunction(new BoaAny(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaTypeVar("V") }, "${0}.remove(${1})"));
+		globalFunctions.addFunction("removeAll", new BoaFunction(new BoaAny(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaSet(new BoaTypeVar("V")) }, "${0}.removeAll(${1})"));
```
Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: this is difference() below?

File: src/java/boa/compiler/SymbolTable.java [hunk 4]
```diff
@@ -188,7 +193,10 @@
 		// set functions
 		globalFunctions.addFunction("contains", new BoaFunction(new BoaBool(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaTypeVar("V") }, "${0}.contains(${1})"));
 		globalFunctions.addFunction("add", new BoaFunction(new BoaAny(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaTypeVar("V") }, "${0}.add(${1})"));
+		globalFunctions.addFunction("addAll", new BoaFunction(new BoaAny(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaSet(new BoaTypeVar("V")) }, "${0}.addAll(${1})"));
+		globalFunctions.addFunction("containsAll", new BoaFunction(new BoaBool(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaSet(new BoaTypeVar("V"))}, "${0}.containsAll(${1})"));
```
Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: is this needed? same as doing: intersect(a, b) == len(b)

Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: I think having familiar names would be helpful.

File: src/java/boa/compiler/SymbolTable.java (hunk 3 above)
Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: I think having familiar names would be helpful.

File: src/java/boa/compiler/SymbolTable.java (hunk 2 above)
Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: Does addAll work for sets only, or is it intended to work for collections also?

File: src/java/boa/compiler/SymbolTable.java (hunk 4 above)
Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: To be precise, some syntactic sugars would be helpful for users.

File: src/java/boa/compiler/SymbolTable.java (hunk 3 above)
Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: Another related question, is removeAll meant to be a mutable operation?

File: src/java/boa/compiler/SymbolTable.java (hunk 2 above)
Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: This is for sets only.

Boa has no general notion of 'collection' as we don't have inheritance.

Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: I'm a fan of union() because it does not mutate either argument.  You are given their union and have to store it somewhere.  Also as far as names go, 'union' *is* the general term for this and more widely known.  The 'addAll' is Java specific.

Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: Thanks for clarifying. Is addAll meant to be a mutable operation?

File: src/java/boa/compiler/SymbolTable.java (hunk 3 above)
Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: I'm a fan of different() because it does not mutate either argument.  You are given their difference and have to store it somewhere.  Also as far as names go, 'difference' *is* the general term for this and more widely known.  The 'removeAll' is Java specific.

**Comment by RamIowaStateUniversity** [comment]
Right now, it is meant to be a mutable operation but as Dr.Dyer said, union
serves the purpose and this is redundant.

File: src/java/boa/compiler/SymbolTable.java (hunk 2 above)
Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: As this is written, it is adding all of 2nd argument into the 1st argument.  This is error prone as people will forget which one is mutated.

Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: All of these good points. Also, consider the performance implication and then make a decision based on that. Thanks both.

File: src/java/boa/compiler/SymbolTable.java (hunk 3 above)
Comment (hridesh) [inline src/java/boa/compiler/SymbolTable.java]: All of these good points. Also, consider the performance implication and then make a decision based on that. Thanks, both.

**Comment by hridesh** [comment]
Ok, changes under discussion have been reverted.
